#!/bin/bash
#PBS -N s1
#PBS -l nodes=1:ppn=10
#PBS -l walltime=480:00:00
#PBS -e errlog.out
#PBS -q batch

echo "Start at"
date

cd $PBS_O_WORKDIR
# bwa mem mapping ;
read1=/public/home/jyli/RawData/Off_Target_Projet/clean_data/s1_H2MHNDMXX_L1_1.clean.fq.gz
read2=/public/home/jyli/RawData/Off_Target_Projet/clean_data/s1_H2MHNDMXX_L1_2.clean.fq.gz
name=s1
bwa mem -M -t 10 -R '@RG\tID:${name}\tSM:${name}_SM' /public/home/jyli/Public_data/HZAU_PacBio_GhV1/SNP_INDEX/Ghirsutum_genome.fasta ${read1} ${read2} >${name}.sam

# sort
java -Xmx20g -jar /public/home/cotton/software/picard-tools/SortSam.jar INPUT=${name}.sam OUTPUT=${name}_srt.bam SORT_ORDER=coordinate >${name}.sam_srt 2>err.${name}.sam_srt

# mark duplicates
java -Xmx20g -jar /public/home/cotton/software/picard-tools/MarkDuplicates.jar INPUT=${name}_srt.bam OUTPUT=${name}_srt_redup.bam METRICS_FILE=metrics.txt > ${name}_redup 2>err.${name}_redup

# index
java -Xmx20g -jar /public/home/cotton/software/picard-tools/BuildBamIndex.jar INPUT=${name}_srt_redup.bam >${name}_index 2>${name}_index

# target regions realignment
/usr/bin/java -Xmx20g -jar /public/home/cotton/software/GATK-3.1.1/GenomeAnalysisTK.jar -T RealignerTargetCreator -R /public/home/jyli/Public_data/HZAU_PacBio_GhV1/SNP_INDEX/Ghirsutum_genome.fasta -I ${name}_srt_redup.bam -o ${name}_forIndelRealigner.intervals -nt 10 -allowPotentiallyMisencodedQuals

/usr/bin/java -Xmx20g -jar /public/home/cotton/software/GATK-3.1.1/GenomeAnalysisTK.jar -T IndelRealigner -R /public/home/jyli/Public_data/HZAU_PacBio_GhV1/SNP_INDEX/Ghirsutum_genome.fasta -I ${name}_srt_redup.bam -targetIntervals ${name}_forIndelRealigner.intervals -o ${name}_realigned.bam -allowPotentiallyMisencodedQuals

# gatk SNP calling
/usr/bin/java -Xmx30g -jar /public/home/cotton/software/GATK-3.1.1/GenomeAnalysisTK.jar -allowPotentiallyMisencodedQuals -R /public/home/jyli/Public_data/HZAU_PacBio_GhV1/SNP_INDEX/Ghirsutum_genome.fasta -T UnifiedGenotyper -glm snp -I ${name}_srt_redup.bam -o ${name}.gatk.snp.vcf -nt 10 -stand_call_conf 30.0 -stand_emit_conf 0

#samtools SNP calling
/public/home/cotton/software/samtools/samtools mpileup -DSugf /public/home/jyli/Public_data/HZAU_PacBio_GhV1/SNP_INDEX/Ghirsutum_genome.fasta ${name}_srt_redup.bam |/public/home/cotton/software/samtools/bcftools/bcftools view -Ncvg - >${name}.samtools.vcf

#indel calling
/usr/bin/java -Xmx30g -jar /public/home/cotton/software/GATK-3.1.1/GenomeAnalysisTK.jar -allowPotentiallyMisencodedQuals -R /public/home/jyli/Public_data/HZAU_PacBio_GhV1/SNP_INDEX/Ghirsutum_genome.fasta -T UnifiedGenotyper -glm INDEL -o ${name}.gatk.indel.vcf -nt 10 -rf BadCigar -I ${name}_realigned.bam -metrics ${name}.gatk.indel.metrics
rm -f *.sam *._srt.bam
echo "End at:"
date
